### ISSUE TITLE
`TypeGuard` error raised on L003 check

### ISSUE THREAD

**Issue by edgarrmondragon** [sqlfluff/sqlfluff#3044]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

With release `0.12.0` I started seeing this error when linting

```﻿﻿
﻿Traceback (most recent call last):                                                                   
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/core/rules/base.py", line 539, in crawl
    res = self._eval(context=context)
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 428, in _eval
    result = self._process_working_state(memory, trigger_segment)
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 483, in _process_working_state
    return self._handle_template_blocks(
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 771, in _handle_template_blocks
    assert template_line, "TypeGuard"
AssertionError: TypeGuard
```

### Expected Behaviour

`L003` check should fail or pass without this unexpected exception.

### Observed Behaviour

Complete output:

```console
$ poetry run sqlfluff lint models
=== [dbt templater] Sorting Nodes...                                                                 
=== [dbt templater] Compiling dbt project...                                                         
=== [dbt templater] Project Compiled.                                                                
CRITICAL   [L003] Applying rule L003 threw an Exception: TypeGuard                                   
Traceback (most recent call last):                                                                   
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/core/rules/base.py", line 539, in crawl
    res = self._eval(context=context)
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 428, in _eval
    result = self._process_working_state(memory, trigger_segment)
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 483, in _process_working_state
    return self._handle_template_blocks(
  File "/Users/edgarramirez/Library/Caches/pypoetry/virtualenvs/dbt-sqlite-test-YBloR8cG-py3.9/lib/python3.9/site-packages/sqlfluff/rules/L003.py", line 771, in _handle_template_blocks
    assert template_line, "TypeGuard"
AssertionError: TypeGuard
== [models/retail/cohorts.sql] FAIL                                                                  
L:   8 | P:  41 | L003 | Unexpected exception: TypeGuard; Could you open an issue                    
                       | at https://github.com/sqlfluff/sqlfluff/issues ? You can
                       | ignore this exception for now, by adding '-- noqa: L003'
                       | at the end of line 8
All Finished 📜 🎉!
```

### How to reproduce

Templated SQL:

```sql
select
  cohort_month
  {% for i in range(15) -%}
  , cast(
    round(sum(case when cohort_index = {{ i }} then amount else 0 end), 2) as real
  ) as index_{{ i }}
  {% endfor -%}
  , cast(sum(amount) as real) as overall
from {{ ref('orders') }}
group by cohort_month
```

<details>
  <summary>Compiled SQL:</summary>

  ```sql
select
  cohort_month
  , cast(
    round(sum(case when cohort_index = 0 then amount else 0 end), 2) as real
  ) as index_0
  , cast(
    round(sum(case when cohort_index = 1 then amount else 0 end), 2) as real
  ) as index_1
  , cast(
    round(sum(case when cohort_index = 2 then amount else 0 end), 2) as real
  ) as index_2
  , cast(
    round(sum(case when cohort_index = 3 then amount else 0 end), 2) as real
  ) as index_3
  , cast(
    round(sum(case when cohort_index = 4 then amount else 0 end), 2) as real
  ) as index_4
  , cast(
    round(sum(case when cohort_index = 5 then amount else 0 end), 2) as real
  ) as index_5
  , cast(
    round(sum(case when cohort_index = 6 then amount else 0 end), 2) as real
  ) as index_6
  , cast(
    round(sum(case when cohort_index = 7 then amount else 0 end), 2) as real
  ) as index_7
  , cast(
    round(sum(case when cohort_index = 8 then amount else 0 end), 2) as real
  ) as index_8
  , cast(
    round(sum(case when cohort_index = 9 then amount else 0 end), 2) as real
  ) as index_9
  , cast(
    round(sum(case when cohort_index = 10 then amount else 0 end), 2) as real
  ) as index_10
  , cast(
    round(sum(case when cohort_index = 11 then amount else 0 end), 2) as real
  ) as index_11
  , cast(
    round(sum(case when cohort_index = 12 then amount else 0 end), 2) as real
  ) as index_12
  , cast(
    round(sum(case when cohort_index = 13 then amount else 0 end), 2) as real
  ) as index_13
  , cast(
    round(sum(case when cohort_index = 14 then amount else 0 end), 2) as real
  ) as index_14
  , cast(sum(amount) as real) as overall
from main."orders"
group by cohort_month
```

</details>

It's a public project too: https://github.com/edgarrmondragon/dbt-sqlite-example/pull/46 🙂 

### Dialect

SQLite

### Version

* SQLFluff: 0.12.0
* Python: 3.9
* `sqlfluff-templater-dbt`: 0.12.0
* dbt: 1.0.4

### Configuration

```toml
# pyproject.toml
[tool.sqlfluff.core]
dialect = "sqlite"
templater = "dbt"
exclude_rules = "L005" # Allow commas to have whitespace directly before them.

[tool.sqlfluff.rules]
tab_space_size = 2
comma_style = "leading"
max_line_length = 100
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by OTooleMichael** [comment]
Interestingly the erorr here is due to the jinja swallow/strip whitespace character

```sql
SELECT
        cohort_month
        {% for i in range(3) -%}
            , {{ i }} AS index_{{ i }}
        {% endfor -%}
        , TRUE AS overall
    FROM orders
```
This is messing up the pairing of open and close somehow`-%}`

**Comment by OTooleMichael** [comment]
Fix  #3065
